### PR TITLE
Sample: App Notifications C++ WinUI Unpackaged

### PR DESCRIPTION
**Pull request by loneursid**
## Description

App Notifications C++ WinIU unpackaged sample 

## Target Release

WindowsAppSDK v1.1

## Checklist

- [ ] Samples build and run using the Visual Studio versions listed in the [Windows development docs](https://docs.microsoft.com/windows/apps/windows-app-sdk/set-up-your-development-environment?tabs=stable#2-install-visual-studio).
- [ ] Samples build and run on all supported platforms (x64, x86, ARM64) and configurations (Debug, Release).
- [ ] Samples set the minimum supported OS version to Windows 10 version 1809.
- [ ] Samples build clean with no warnings or errors.
- [ ] **[For new samples]**: Samples have completed the [sample guidelines checklist](https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/docs/samples-guidelines.md#checklist) and follow [standardization/naming guidelines](https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/docs/samples-guidelines.md#standardization-and-naming).
- [ ] If I am onboarding a new feature, then I must have correctly setup a new CI pipeline for my feature with the correct triggers and path filters laid out in the "Onboarding Samples CI Pipeline for new feature" section in samples-guidelines.md.
- [ ] I have commented on my PR `/azp run` to have the CI build run on my branch with the right feature pipeline. This must be done right before merging to ensure the build is up to date and accurate.

### CODE REVIEW

File: Samples/nuget.config [hunk 1]
```diff
@@ -10,6 +10,7 @@
     <packageSources>
     <clear />
         <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
+        <add key="Local" value="C:\Users\erlangl\Downloads" />
```
Comment (erik0686) [inline Samples/nuget.config:13]: I see this is still a draft, so maybe I'm getting ahead of myself, but we might want to remove this :)

Comment (loneursid) [reply Samples/nuget.config:13]: We will, as soon as we have WindowsAppSDK v1.1 nugets available on a feed. 

This sample is in it's very beginning and still has a long way to go!

Thanks for looking...

Comment (erik0686) [reply Samples/nuget.config:13]: Ah I see, thank you for this sample!

File: Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/ToastWithTextBox.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    using namespace CppUnpackagedAppNotifications::implementation;
+}
+
+bool ToastWithTextBox::SendToast()
+{
+    winrt::hstring xmlPayload{
+        L"<toast>\
+                <visual>\
+                    <binding template = \"ToastGeneric\">\
+                        <image placement = \"appLogoOverride\" src = \"" + winrt::App::GetFullPathToAsset(L"Square150x150Logo.png") + L"\"/>\
```
[the remark this answers is not in the document]
Comment (loneursid) [reply Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/ToastWithTextBox.cpp]: Done. My only concern is with "Open App". We don't always open the app.

[the remark this answers is not in the document]
Comment (vaheeshta) [reply Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/ToastWithTextBox.cpp]: Can you help me understand why that is?

I noticed when the app is minimized or not in the foreground, selecting the notification doesn't bring the app into the foreground. This goes against this guideline we've added to the quickstart:

image

If @hulumane is in agreement, could we please implement the above guideline for when the notification body is selected? 

I also noticed your notifications do not have a launch args in the <toast> element. Is there a reason why that is? Ideally we'd want to add lanch args in the <toast> element to 1) give developers an example of how to process launch args in a toast without buttons and 2) selecting the toast body in this app gives the "Unrecognized Toast Originator" message.

Let me know if I'm missing something in my understanding.

[the remark this answers is not in the document]
Comment (loneursid) [reply Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/ToastWithTextBox.cpp]: *I noticed when the app is minimized or not in the foreground, selecting the notification doesn't bring the app into the foreground.*
That's on my todo list for Friday 

*I also noticed your notifications do not have a launch args in the <toast> element.*
That's on my to-do for Thursday

[the remark this answers is not in the document]
Comment (loneursid) [reply Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/ToastWithTextBox.cpp]: Fixed

**Review by danielayala94 — APPROVED** [review]

**Comment by danielayala94** [comment]
Reviewed offline with Eric. Some nits but not any general concerns. This is good to go.

:shipit:

File: Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    using namespace Microsoft::UI::Xaml;
+    using namespace winrt::Microsoft::Windows::AppLifecycle;
+    using namespace winrt::Microsoft::Windows::AppNotifications;
+}
+
+// NotificationManager is responsible for registering and unregistering the Sample for App Notifications as well as
+// dispatching actionned notifications to the appropriate scenario.
+// Registration will happen when Init() is called and Unregistration will happen when this
+// instance variable goes out of scope, i.e.: when the App is terminated.
+static NotificationManager g_notificationManager;
```
Comment (hulumane) [inline Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp:25]: Nit: A better solution would be to have a API called GetNotificationManager() which returns a manager instance. The manager in the APIs context would be initialized as static.

File: Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        HWND hwnd;
+        auto WindowNative{ app->window.as<IWindowNative>() };
+        if (WindowNative != nullptr && WindowNative->get_WindowHandle(&hwnd) == S_OK)
+        {
+            SwitchToThisWindow(hwnd, TRUE);
+        }
+    }
+
+    std::wstring App::GetFullPathToExe()
```
Comment (hulumane) [inline Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp:60]: Good place to start using r-value refs. So this method can return std::wstring&& and the caller could use std::move for assignment which would avoid a copy.

Comment (hulumane) [reply Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp:60]: In general, we could do this everywhere we utilize and return strings.

File: Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        // NOTE: AppInstance is ambiguous between
+        // Microsoft.Windows.AppLifecycle.AppInstance and
+        // Windows.ApplicationModel.AppInstance
+        auto currentInstance{ winrt::AppInstance::GetCurrent() };
+        if (currentInstance)
+        {
+            // AppInstance.GetActivatedEventArgs will report the correct ActivationKind,
+            // even in WinUI's OnLaunched.
+            winrt::AppActivationArguments activationArgs{ currentInstance.GetActivatedEventArgs() };
+            if (activationArgs)
```
Comment (hulumane) [inline Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.cpp:88]: Is this check needed. I think we are guaranteed an activationArgs of Launch at the minimal every single time the app launches.

File: Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+#include "App.xaml.g.h"
+#include "pch.h"
+
+namespace winrt::CppUnpackagedAppNotifications::implementation
+{
+    struct App : AppT<App>
+    {
+        App();
+
+        static std::wstring GetFullPathToExe();
```
Comment (hulumane) [inline Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/App.xaml.h:14]: Should this be private?

File: Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/Notifications/ToastWithAvatar.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+{
+    // The ScenarioIdToken uniquely identify a scenario and is used to route the response received when the user clicks on a toast to the correct scenario.
+    auto ScenarioIdToken{ Common::MakeScenarioIdToken(ToastWithAvatar::ScenarioId) };
+
+    winrt::hstring xmlPayload{
+        L"<toast launch = \"action=ToastClick&amp;" + ScenarioIdToken + L"\">\
+            <visual>\
+                <binding template = \"ToastGeneric\">\
+                    <image placement = \"appLogoOverride\" hint-crop=\"circle\" src = \"" + winrt::App::GetFullPathToAsset(L"Square150x150Logo.png") + L"\"/>\
+                    <text>" + ScenarioName + L"</text>\
```
Comment (hulumane) [inline Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/Notifications/ToastWithAvatar.cpp:29]: You can avoid a lot of the L" and double "" by doing this:

winrt::hstring payload =
        LR"(<toast launch="action = viewDownload &amp; downloadId = 9438108">
        <visual>
            <binding template = "ToastGeneric">
                <text>Downloading this week's new music...</text>
                <progress
                    title = "{progressTitle}"
                    value = "{progressValue}"
                    valueStringOverride = "{progressValueString}"
hulumane marked this conversation as resolved.
hulumane marked this conversation as resolved.
                    status = "{progressStatus}" />
            </binding>
        </visual>
    </toast>)";

Comment (hulumane) [reply Samples/Notifications/App/CppUnpackagedAppNotifications/CppUnpackagedAppNotifications/Notifications/ToastWithAvatar.cpp:29]: https://github.com/microsoft/WindowsAppSDK/pull/1675/files#:~:text=winrt%3A%3Ahstring%20payload,%3C/toast%3E)%22%3B

**Pull request merged**